owncloud/web
# [full-ci] feat: introduce contrast colors in swatches

## Description
This PR introduces a `contrast` token for all our `swatches` (i.e. our color variations in the design system). Their purpose is to define a high contrast color in relation to the `default` token of the swatches. This way the `contrast` token can e.g. be used as text color if the `default` token represents a background color.

The `contrast` tokens are mainly used in three components of the design system:
- OcModal
- OcButton
- OcIcon
The main difference in those components now is that they don't use a generic text color anymore, but instead use the `contrast` color of the active variation. This brings us the flexibility of having light and dark swatches in parallel. Previously this was not possible because one single text color was used for all swatches/variations. Having a light brand color and a dark `danger` color was conflicting. One of the two always had bad readability for text.

One additional goal was to completely get rid of the `inverse` swatch. Because, come on, `inverse` to what exactly?! I'm not there, yet, there are still 4 usages of the `inverse` swatch in the code. Might push that to a followup.

## Motivation and Context
More flexibility in theming.

## How Has This Been Tested?
Well, it's about colors. Manually.
Mainly there are two areas of interest:
- check that the existing `owncloud` default theme looks the same as before with this PR.
- try out your own custom theme with a light `brand` color. E.g. the json at the end of this PR description.

## Types of changes


## Checklist:




## Example theme.json with light brand color

In order to test this, create a new theme folder in `packages/web-runtime/themes`, insert the following content in a `theme.json` file. Add logos as needed in an assets folder. Build web and set the theme path in the `docker-compose.yaml` file.

```json
{
	"common": {
		"name": "Custom",
		"slogan": "Custom - Your custom cloud",
		"logo": "themes/custom/assets/logo.svg"
	},
	"ios": {},
	"web": {
		"default": {
			"general": {
				"name": "Custom",
				"slogan": "Custom - Your custom cloud"
			},
			"logo": {
				"topbar": "themes/custom/assets/logo.svg",
				"favicon": "themes/custom/assets/favicon.ico",
				"login": "themes/custom/assets/logo.svg"
			},
			"loginPage": {
				"autoRedirect": true,
				"backgroundImg": "themes/custom/assets/loginBackground.svg"
			},
			"designTokens": {
				"colorPalette": {
					"background-accentuate": "rgba(255, 255, 5, 0.1)",
					"background-default": "#ffffff",
					"background-highlight": "#edf3fa",
					"background-muted": "#f8f8f8",
					"background-secondary": "#ffffff",
					"background-hover": "rgb(236, 236, 236)",
					"border": "#ecebee",
					"input-bg": "#f0f4f8",
					"input-border": "#ceddee",
					"input-text-default": "#041e42",
					"input-text-muted": "#4c5f79",
					"swatch-brand-default": "rgb(238, 238, 238)",
					"swatch-brand-hover": "#223959",
          "swatch-brand-muted": "#223959",
          "swatch-brand-contrast": "rgb(39, 77, 127)",
					"swatch-danger-default": "#e00000",
					"swatch-danger-hover": "#e96464",
					"swatch-danger-muted": "#b50000",
					"swatch-inverse-default": "#ffffff",
					"swatch-inverse-hover": "#ffffff",
					"swatch-inverse-muted": "#bfbfbf",
					"swatch-passive-default": "#4c5f79",
					"swatch-passive-hover": "#97a2b1",
					"swatch-passive-muted": "#283e5d",
					"swatch-primary-default": "#4a76ac",
					"swatch-primary-hover": "#80a7d7",
					"swatch-primary-muted": "#2c588e",
					"swatch-primary-gradient": "#4e85c8",
					"swatch-success-default": "#4f822f",
					"swatch-success-hover": "#93e054",
					"swatch-success-muted": "#53960a",
					"swatch-warning-default": "#c6501b",
					"swatch-warning-hover": "#fda94b",
					"swatch-warning-muted": "#b35100",
					"text-default": "#041e42",
					"text-inverse": "#ffffff",
					"text-muted": "#4c5f79",
					"icon-folder": "#4d7eaf",
					"icon-archive": "#fbbe54",
					"icon-image": "#ee6b3b",
					"icon-spreadsheet": "#15c286",
					"icon-document": "#3b44a6",
					"icon-video": "#045459",
					"icon-audio": "#700460",
					"icon-presentation": "#ee6b3b",
					"icon-pdf": "#ec0d47"
				}
			}
		}
	}
}
```
